### ISSUE TITLE
compareTo() not being invoked when using an interface as type parameter for List<T>

### ISSUE THREAD

**Issue by hvaldecantos** [jcoplien/trygve#97]
``` java
interface Countable {
    public int value() const;
    public int compareTo(Countable other) const;
}
class Number implements Countable {
    int number_;
    public Number(int num) { number_ = num.clone; }
    public int value() const { return number_; }
    public int compareTo(Countable other) const { return value() - other.value(); }
}
{
    Countable one = new Number(1)
    Countable another_one = new Number(1)

    List<Countable> countables = new List<Countable>()
    countables.add(one)

    System.out.println(countables.contains(one))
    System.out.println(countables.contains(another_one))
    System.out.println(countables.contains(new Number(1))) // [*]
}
```

The execution using trygve 1.6.9 outputs:

```
Compilation complete
true
false
false
```

The programmer-supplied compareTo method is not being invoked. This seems to be related to issue  #91

---

`[*]` Not sure but, shouldn't the compiler say something like:

```
line 20: Script `contains(Number)' not declared in class List<Countable>
```

as it would do if we try to run `one.compareTo(new Number(1))`?

**Comment by jcoplien** [comment]
Interesting. There is no reason this shouldn’t work. I somehow suspect it relates to the use of the interface. A few quick investigations didn’t turn up anything so I’m going to have to dig into it.

Thanks for reporting it.

First bug in a week ;-)

**Comment by jcoplien** [comment]
I think this is fixed, though your code is a bit roundabout.  My guess is that it would have worked had compareTo taken a Number argument.

**Comment by hvaldecantos** [comment]
Yes, it did work if `compareTo` took a `Number` argument, but  I wouldn't have been using the interface. Although not needed for this simplified example, I was using the `compareTo()` within a composite structure for a menu to check that a menu option (menu item or sub-menu) was not repeated. I used the composite pattern to define the class structure, whereas the interactions in a menu interface context.

I see that there's still some issues regarding relational operators.

``` java
{
    ...
    System.out.println(one == another_one)
    System.out.println(one < another_one)
    System.out.println(one > another_one)
    System.out.println(one <= another_one)
    System.out.println(one >= another_one)
    System.out.println(one != another_one)
}
```

```
line 21: To use `==' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
line 22: To use `<' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
line 23: To use `>' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
line 24: To use `<=' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
line 25: To use `>=' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
line 26: To use `!=' on object `one' you must declare a compareTo(Countable argument) operation in the `requires' section of Countable
Compilation complete
6 fatal errors
```

Note the compiler message, it is saying something about the  `requires' section of Countable, when Countable is an interface. I haven't checked these issues with other generic classes, anyhow I don't plan to use them for the time being.

**Comment by jcoplien** [comment]
I think this is fixed.

**Comment by hvaldecantos** [comment]
Yes, it is fixed since version 1.6.12.
Thanks!

**Comment by jcoplien** [comment]
Thanks Héctor!
